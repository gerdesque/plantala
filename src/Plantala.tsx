import React from 'react';
import Header from './header/Header';
import './Plantala.css';

function Plantala() {
  return (
    <div className="plantala">
      <Header />
    </div>
  );
}

export default Plantala;
