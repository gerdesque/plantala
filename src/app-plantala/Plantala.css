.plantala {
  height: 100vh;
  width: 100vw;
  display: grid;
  grid-template-rows: auto 1fr auto;
}

@media print {
  html, body { width: 210mm; height: 297mm;}
  footer, img, button, h5 { display: none !important; }
  .MuiAppBar-colorPrimary {box-shadow: none !important;}
  h2::before { content: "Malen. Strahlen. "; }
  h2::after { content: "!"; }
  .MuiContainer-root { max-width: 1600px !important; padding: 0 !important }
}