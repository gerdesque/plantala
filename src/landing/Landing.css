.app-landing {
  width: 40vmin;
  height: auto;
  pointer-events: none;
}

@media (prefers-reduced-motion: no-preference) {
  .app-landing {
    animation: app-landing-spin infinite 20s linear;
  }
}

@keyframes app-landing-spin {
  from {
    transform: rotate(0deg);
  }
  to {
    transform: rotate(360deg);
  }
}
