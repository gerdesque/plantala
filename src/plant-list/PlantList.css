.image-item {
  padding: 2em 0;
  filter: brightness(.9);
}

.react-multi-carousel-item--active {
  padding: 1em 0;
  filter: brightness(1);
  transform: perspective(800px) translate3d(0, 0, 25px);
}
