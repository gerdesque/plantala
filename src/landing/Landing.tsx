import React from 'react';
import { Grid } from '@material-ui/core';
import logo from '../assets/logo.webp';
import './Landing.css';

export default function Landing() {
  return (
    <Grid className="app-landing-container" container justify="center">
      <img src={logo} className="app-landing" alt="Plantala" />
    </Grid>
  );
}
