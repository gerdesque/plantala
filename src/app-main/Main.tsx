import React from 'react';
import { Button, Container, Grid, Typography } from '@material-ui/core';
import { makeStyles } from '@material-ui/core/styles';
import { IPlant } from '../plant/Plant';
import Landing from '../landing/Landing';
import PlantList from '../plant-list/PlantList';
import { Action } from '../app-plantala/Plantala';

export enum ButtonValue {
  Start = "Start",
  Select = "Zur Auswahl hinzufügen",
  Done = "Fertig",
  Again = "Neues Plantala erstellen"
}

export enum HeaderValue {
  Start = "Erstelle dein Mandala!",
  Select = "Wähle bis zu 5 Karten aus",
  Done = "Gestalte dein Plantala",
  Again = "Teile dein Plantala"
}

interface IMainProps {
  plants: IPlant[],
  action: Action,
  setAction: any
}

const useStyles = makeStyles((theme) => ({
  heroContent: {
    backgroundColor: theme.palette.background.default,
    padding: theme.spacing(2),
  },
  content: {
    paddingTop: theme.spacing(4),
  },
  main: {
    backgroundColor: theme.palette.primary.main,
  },
}));

export default function Main({ plants, action, setAction }: IMainProps) {
  const classes = useStyles();

  return (
    <main className={classes.main}>
      <Container maxWidth="md" className={classes.heroContent}>
        <Typography variant="h4" align="center" color="textPrimary" gutterBottom>
          {HeaderValue[action]}
        </Typography>
        <div>
          <Grid container spacing={2} justify="center">
            <Grid item md={3}>
              <Button variant="contained" color="secondary" fullWidth onClick={setAction}>
                {ButtonValue[action]}
              </Button>
            </Grid>
          </Grid>
        </div>
      </Container>
      <Container className={classes.content} maxWidth="md">
        {action === Action.Start && <Landing /> }
        {action === Action.Select && <PlantList plants={plants} /> }
          {/* <Mandala /> */}
      </Container>
    </main>
  );
}
