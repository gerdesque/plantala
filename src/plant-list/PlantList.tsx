import React, {useContext, useState} from 'react';
import { Button, Card, CardActions, CardContent, CardHeader, Avatar } from '@material-ui/core';
import { makeStyles } from '@material-ui/core/styles';
import Carousel from "react-multi-carousel";
import "react-multi-carousel/lib/styles.css";
import './PlantList.css';
import { IPlant } from '../plant/Plant';
import LazyCardMedia from './LazyCardMedia';
import Description from './Description';
import { avatarCount } from '../utils/Utils';
import { AppContext } from '../app-plantala/Context';

interface IPlantListProps {
  plants: IPlant[],
  setSelectedPlant: any,
  selectedPlants: IPlant[],
}

const useStyles = makeStyles((theme) => ({
  plant: {
    height: '100%',
    display: 'flex',
    flexDirection: 'column',
    textAlign: 'center',
    borderColor: theme.palette.primary.main,
    borderStyle: 'solid',
    borderWidth: theme.spacing(1),
    cursor: 'pointer'
  },
  selected: {
    borderColor: theme.palette.secondary.main,
  },
  plantHeader: {
    backgroundColor: theme.palette.primary.light,
    borderColor: theme.palette.common.black,
    borderStyle: 'solid',
    borderWidth: theme.spacing(0.5),
    marginLeft: theme.spacing(1),
    marginRight: theme.spacing(1),
    marginTop: theme.spacing(1),
  },
  plantMedia: {
    paddingTop: '100%',
  },
  plantAvatar: {
    width: theme.spacing(12),
    height: theme.spacing(12),
    position: 'absolute',
    top: '50%',
    left: '50%',
    marginTop: -theme.spacing(6),
    marginLeft: -theme.spacing(6),
  },
  plantContent: {
    backgroundColor: theme.palette.primary.light,
    borderColor: theme.palette.common.black,
    borderStyle: 'solid',
    borderWidth: theme.spacing(0.5),
    marginLeft: theme.spacing(1),
    marginRight: theme.spacing(1),
    marginTop: theme.spacing(1),
    padding: theme.spacing(1),
    hyphens: 'auto',
  },
}));

export default function PlantList({ plants, setSelectedPlant, selectedPlants }: IPlantListProps) {
  const classes = useStyles();
  const responsive = {
    desktop: {
      breakpoint: { max: 5000, min: 1024 },
      items: 1,
      partialVisibilityGutter: 10
    },
    tablet: {
      breakpoint: { max: 1024, min: 464 },
      items: 1
    },
    mobile: {
      breakpoint: { max: 464, min: 0 },
      items: 1
    }
  };
  const [isMoving, setMoving] = useState(false);
  const isPlantListFull = selectedPlants.length === avatarCount;
  const colorMode = useContext(AppContext);
  const colorPath = colorMode ? '_Bunt' : '_SW';

  const [fadePlant, setFadingPlant] = useState({} as IPlant);
  function selectCard(plant: IPlant):void {
    if (plant.selected) {
      setSelectedPlant(plant);
    } else if (!isPlantListFull) {
      setFadingPlant(plant);
      setTimeout(function () {
        setFadingPlant(plant)
        setSelectedPlant(plant)
      }, 500);
    }
  }

  return (
      <Carousel
      itemClass="image-item"
      responsive={responsive}
      renderDotsOutside
      infinite={true}
      autoPlay={false}
      containerClass="carousel-container"
      customTransition="all .3s ease-in-out"
      transitionDuration={300}
      centerMode={true}
      removeArrowOnDeviceType={["tablet", "mobile"]}
      beforeChange={() => setMoving(true) }
      afterChange={() => setMoving(false) }
    >
      {plants.map(plant => (
        <Card 
          key={plant.name}
          className={`${classes.plant} ${plant.selected ? classes.selected : ''}`}
          onClick={() => {
            if (!isMoving) {
              selectCard(plant)
            }
            }}>
          <CardHeader
            title={plant.name}
            className={classes.plantHeader}
          />
          <LazyCardMedia
            className={classes.plantMedia}
            image={require('../assets/' + plant.source + colorPath + '.png').default}
            title={plant.name}
          />
          {fadePlant === plant && <Avatar
            key={plant.name}
            alt={plant.name}
            src={require('../assets/' + plant.source + colorPath + '.png').default}
            className={`${classes.plantAvatar} ${fadePlant === plant ? 'fade' : ''}`}
          />}
          {plant.description && <CardContent className={classes.plantContent}>
            <Description description={plant.description} />
          </CardContent>}
          <CardActions>
            <Button size="small" color="secondary">
              Mehr erfahren
            </Button>
          </CardActions>
        </Card>
      ))}
      </Carousel>
  );
}
