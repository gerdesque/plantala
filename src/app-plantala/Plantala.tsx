import React, { Component } from 'react';
import Header from '../app-header/Header';
import Main from '../app-main/Main';
import Footer from '../app-footer/Footer';
import './Plantala.css';
import { plants } from '../plants';
import { IPlant } from '../plant/Plant';

export enum Action {
  Start = "Start",
  Select = "Select",
  Done = "Done",
  Again = "Again"
}

interface IPlantalaState {
  plants: IPlant[],
  selectedPlants: IPlant[],
  activePlant: IPlant,
  action: Action
}

class Plantala extends Component {
  state: IPlantalaState = {
    // values to be displayed in <Card />
    plants: plants,
    // plants selected for <AvatarGroup />
    selectedPlants: plants,
    // active plant for <Plant />
    activePlant: {} as IPlant,
    // action mode to be displayed in <Main />
    action: Action.Start,
  }

  setAction = () => {
    this.setState({action: Action.Select});
  }

  render = () => {
    const { plants, selectedPlants, action } = this.state;

    return (
      <div className="plantala">
        <Header />
        <Main
          plants={plants}
          action={action}
          setAction={this.setAction} />
        <Footer
          selectedPlants={selectedPlants} />
      </div>
    );
  }
}

export default Plantala;
