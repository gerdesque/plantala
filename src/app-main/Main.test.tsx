import React from 'react';
import { Button, Typography } from '@material-ui/core';
import { createShallow } from '@material-ui/core/test-utils';
import { ShallowWrapper } from 'enzyme';
import Main from './Main';
import { ButtonValue, HeaderValue } from '../app-plantala/Plantala';
import Landing from '../landing/Landing';
import PlantList from '../plant-list/PlantList';

describe('Main in Start mode', () => {

  let shallow: any;
  let wrapper: ShallowWrapper<any, Readonly<{}>, React.Component<{}, {}, any>>;

  beforeEach(() => {
    shallow = createShallow();
    wrapper = shallow(<Main
      plants={[]}
      displayedButtonValue={ButtonValue.Start}
      displayedHeaderValue={HeaderValue.Start} />);
  });

  it('should render correctly', () => expect(wrapper).toMatchSnapshot());

  it('should display the text "Erstelle dein Mandala!"', () => {
    expect(wrapper.find(Typography).text()).toEqual('Erstelle dein Mandala!');
  });

  it('should display a <Button /> with the text "Start"', () => {
    expect(wrapper.find(Button).text()).toEqual('Start');
  });

  it('should display a <Landing /> component', () => {
    expect(wrapper.containsMatchingElement(<Landing />)).toEqual(true);
  });

  it('should display a <PlantList /> component', () => {
    expect(wrapper.containsMatchingElement(<PlantList plants={[]} />)).toEqual(false);
  });

});

describe('Main in Selection mode', () => {

  let shallow: any;
  let wrapper: ShallowWrapper<any, Readonly<{}>, React.Component<{}, {}, any>>;

  beforeEach(() => {
    shallow = createShallow();
    wrapper = shallow(<Main
      plants={[]}
      displayedButtonValue={ButtonValue.Select}
      displayedHeaderValue={HeaderValue.Select} />);
  });

  it('should render correctly', () => expect(wrapper).toMatchSnapshot());

  it('should display the text "W채hle bis zu 5 Karten aus"', () => {
    expect(wrapper.find(Typography).text()).toEqual('W채hle bis zu 5 Karten aus');
  });

  it('should display a <Button /> with the text "Zur Auswahl hinzuf체gen"', () => {
    expect(wrapper.find(Button).text()).toEqual('Zur Auswahl hinzuf체gen');
  });

  it('should display a <Landing /> component', () => {
    expect(wrapper.containsMatchingElement(<Landing />)).toEqual(false);
  });

  it('should display a <PlantList /> component', () => {
    expect(wrapper.containsMatchingElement(<PlantList plants={[]} />)).toEqual(true);
  });

});
