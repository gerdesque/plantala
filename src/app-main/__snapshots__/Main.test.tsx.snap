// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Main in Selection mode should render correctly 1`] = `
<main
  className="makeStyles-main-3"
>
  <WithStyles(ForwardRef(Container))
    className="makeStyles-heroContent-1"
    maxWidth="md"
  >
    <WithStyles(ForwardRef(Typography))
      align="center"
      color="textPrimary"
      gutterBottom={true}
      variant="h4"
    >
      Wähle bis zu 5 Karten aus
    </WithStyles(ForwardRef(Typography))>
    <div>
      <WithStyles(ForwardRef(Grid))
        container={true}
        justify="center"
        spacing={2}
      >
        <WithStyles(ForwardRef(Grid))
          item={true}
          md={3}
        >
          <WithStyles(ForwardRef(Button))
            color="secondary"
            fullWidth={true}
            onClick={[MockFunction]}
            variant="contained"
          >
            Zur Auswahl hinzufügen
          </WithStyles(ForwardRef(Button))>
        </WithStyles(ForwardRef(Grid))>
      </WithStyles(ForwardRef(Grid))>
    </div>
  </WithStyles(ForwardRef(Container))>
  <WithStyles(ForwardRef(Container))
    className="makeStyles-content-2"
    maxWidth="md"
  >
    <PlantList
      plants={Array []}
    />
  </WithStyles(ForwardRef(Container))>
</main>
`;

exports[`Main in Start mode should render correctly 1`] = `
<main
  className="makeStyles-main-3"
>
  <WithStyles(ForwardRef(Container))
    className="makeStyles-heroContent-1"
    maxWidth="md"
  >
    <WithStyles(ForwardRef(Typography))
      align="center"
      color="textPrimary"
      gutterBottom={true}
      variant="h4"
    >
      Erstelle dein Mandala!
    </WithStyles(ForwardRef(Typography))>
    <div>
      <WithStyles(ForwardRef(Grid))
        container={true}
        justify="center"
        spacing={2}
      >
        <WithStyles(ForwardRef(Grid))
          item={true}
          md={3}
        >
          <WithStyles(ForwardRef(Button))
            color="secondary"
            fullWidth={true}
            onClick={[MockFunction]}
            variant="contained"
          >
            Start
          </WithStyles(ForwardRef(Button))>
        </WithStyles(ForwardRef(Grid))>
      </WithStyles(ForwardRef(Grid))>
    </div>
  </WithStyles(ForwardRef(Container))>
  <WithStyles(ForwardRef(Container))
    className="makeStyles-content-2"
    maxWidth="md"
  >
    <Landing />
  </WithStyles(ForwardRef(Container))>
</main>
`;
